programs/drift-competitions/src/state/competition.rs

```diff
@@ -1,32 +1,404 @@
+            ErrorCode::InvalidRoundSettlementDetected,
+            "{} competitiors not not settled ({} != {})",
+            self.number_of_competitors - self.number_of_competitors_settled,
+            self.number_of_competitors,
+            self.number_of_competitors_settled
+        )?;
+
+        validate!(
+            self.status == CompetitionRoundStatus::WinnerSettlementComplete,
+            ErrorCode::InvalidRoundSettlementDetected,
+            "Competition status = {:?} (should be WinnerSettlementComplete)",
+            self.status
+        )?;
+
+        Ok(())
+    }
+
+    pub fn validate_round_resolved(&self) -> CompetitionResult {
+        validate!(
+            self.number_of_competitors > 0
+                && self.number_of_competitors == self.number_of_competitors_settled,
+            ErrorCode::CompetitionRoundInSettlementPhase,
+            "Competition round_number={:?} is still ongoing",
+            self.round_number
+        )?;
+
+        Ok(())
+    }
+
+    pub fn validate_competitor_is_winner(&self, competitor: &Competitor) -> CompetitionResult {
+        validate!(
+            self.status == CompetitionRoundStatus::WinnerDrawComplete,
+            ErrorCode::CompetitionWinnerNotDetermined
+        )?;
+
+        // competitor account was settled and set to next round
+        validate!(
+            self.round_number.safe_add(1)? == competitor.competition_round_number,
+            ErrorCode::CompetitorHasWrongRoundNumber
+        )?;
+
+        validate!(
+            self.winning_draw >= competitor.min_draw && self.winning_draw < competitor.max_draw,
+            ErrorCode::CompetitorNotWinner
+        )?;
+
+        Ok(())
+    }
+
+    pub fn competitor_can_be_settled(&self, competitor: &Competitor) -> bool {
+        return self.round_number == competitor.competition_round_number;
+    }
+
+    pub fn settle_competitor(
+        &mut self,
+        competitor: &mut Competitor,
+        user_stats: &UserStats,
+        now: i64,
+    ) -> CompetitionResult {
+        self.validate_round_ready_for_settlement(now)?;
+
+        if !self.competitor_can_be_settled(competitor) {
+            return Ok(()); // gracefully skip/fail
+        }
+
+        // skip unclaimed winners to give active competitors a higher probablity of winning
+        if competitor.unclaimed_winnings == 0 {
+            let round_score = competitor.calculate_round_score(user_stats)?;
+
+            let round_score_capped = if self.max_entries_per_competitor > 0 {
+                round_score.min(self.max_entries_per_competitor.cast()?)
+            } else {
+                round_score
+            };
+
+            let new_total_score_settled = self
+                .total_score_settled
+                .safe_add(round_score_capped.cast()?)?;
+
+            competitor.min_draw = self.total_score_settled;
+            competitor.max_draw = new_total_score_settled;
+            competitor.competition_round_number =
```

Comment:
you always need to update this right? otherwise you brick the competitor account? should write a test for this